### ISSUE TITLE
StochasticDualCoordinateAscentClassifier issue when x86 (32 bits)

### ISSUE THREAD

**Issue by DevLob-zz** [dotnet/machinelearning#1721]
### System information

- **OS version/distro**:
Windows 10
- **.NET Version (eg., dotnet --info)**: 
4.6.1
### Issue

- **What did you do?**
try to use StochasticDualCoordinateAscentClassifier as classifier on My ML application
- **What happened?**
i ran it through 86x and through the following exception

> Managed Debugging Assistant 'PInvokeStackImbalance' 
  Message=Managed Debugging Assistant 'PInvokeStackImbalance' : 'A call to PInvoke function 'Microsoft.ML.CpuMath!Microsoft.ML.Runtime.Internal.CpuMath.Thunk::SumSqU' has unbalanced the stack. This is likely because the managed PInvoke signature does not match the unmanaged target signature. Check that the calling convention and parameters of the PInvoke signature match the target unmanaged signature.'

- **What did you expect?**
it should work normally in 86x as it work fine with 64 and this exception occur only with 86x and i need to run it with 86 

### Source code / logs
```pipeline.Add(CollectionDataSource.Create(noramalTagsTrainingData));
pipeline.Add(new Dictionarizer("Label"));
// Transform any text feature to numeric values
pipeline.Add(new CategoricalOneHotVectorizer(
"fontColor",
"tagText",
"firstWord"
));
pipeline.Add(new ColumnConcatenator(
"Features",
"fontSize",
"isBold",
"isItalic",
"isUnderLine",
"containsDot",
"containsQuestionMark",
"isAllCaps",
"tagText",
"firstWord"
));
pipeline.Add(new StochasticDualCoordinateAscentClassifier
            {
                MaxIterations = 100,
                NumThreads = 7,
                LossFunction = new SmoothedHingeLossSDCAClassificationLossFunction()
            });
pipeline.Add(new PredictedLabelColumnOriginalValueConverter() { PredictedLabelColumn = "PredictedLabel" });

 // The pipeline is trained on the dataset that has been loaded and transformed.
var model = pipeline.Train<NormalTagsModelFeatures, NormalTagsPrediction>();
```

it through the exception at last line within only 86x

any help please
Please paste or attach the code or logs or traces that would be helpful to diagnose the issue you are reporting.

**Comment by shmoradims** [comment]
@CESARDELATORRE could you please provide the data file (noramalTagsTrainingData) for repro?

**Comment by DevLob-zz** [comment]
> @CESARDELATORRE could you please provide the data file (noramalTagsTrainingData) for repro?

this is sample data , the original learning data save in SQL Server then load to my application 

using the following code
`using (reader) //close con on reader.Close
            {
                while (reader.Read())
                {
                    noramalTagsTestData.Add(
                        new NormalTagsModelFeatures()
                        {
                            Label = (string)reader["tagName"],
                            fontSize = (float)(int)reader["fontSize"],
                            isBold = (float)Convert.ToSingle(reader["isBold"]),
                            isItalic = (float)Convert.ToSingle(reader["isItalic"]),
                            isUnderLine = (float)Convert.ToSingle(reader["isUnderLine"]),
                            containsDot = (float)Convert.ToSingle(reader["containsDot"]),
                            containsQuestionMark = (float)Convert.ToSingle(reader["containsQuestionMark"]),
                            isAllCaps = (float)Convert.ToSingle(reader["isAllCaps"]),
                            fontColor = (string)reader["fontColor"],
                            tagText = (string)reader["tagText"],
                            firstWord = (string)reader["firstWord"]
                        }
                    );`
[Untitled spreadsheet.xlsx](https://github.com/dotnet/machinelearning/files/2618650/Untitled.spreadsheet.xlsx)


* the same code work fine on 64 Bit and with another algorithm NaiveBayesClassifier on two 32 and 64 bit but  StochasticDualCoordinateAscentClassifier  work only within 64

best regards

**Comment by jasonxz** [comment]
I'm having this exact issue with ALL of the Trainers with the exception of FastTree and FastTreeTweedie.

**Comment by Ivanidzo4ka** [comment]
I push changes which fix problem and they should be available in daily nuget. If you can't use daily nuget i suggest you to disable that exception tracking in your application:
CTRL + ALT + E
Under "Managed Debugging Assistants" uncheck PInvokeStackImbalance.

**Comment by DevLob-zz** [comment]
> I push changes which fix problem and they should be available in daily nuget. If you can't use daily nuget i suggest you to disable that exception tracking in your application:
> CTRL + ALT + E
> Under "Managed Debugging Assistants" uncheck PInvokeStackImbalance.

If i disabled this exception this mean the code will resume his job fine ?

**Comment by Ivanidzo4ka** [comment]
>If i disabled this exception this mean the code will resume his job fine ?

During my investigation that was the case.
Also I think switch for Debug to Release configuration will also allow you to continue training.

**Comment by Ivanidzo4ka** [comment]
Do you have any CPU/Memory consumption during training?
How many examples do you have? 
You can always set 1 iteration for training, that way it would run only one training iteration.

**Comment by DevLob-zz** [comment]
`   public void BuildNormalTrainEvaluateAndSaveModel()
        {
            //Set a random seed for repeatable/deterministic results across multiple trainings.
            var mlContext = new MLContext(seed: 1);

            #region "STEP 1: Common data loading configuration"
            IDataView trainingDataView = GetNormalDataSet(mlContext);

            //var env = mlContext.Data.GetEnvironment();
            //IDataView trainingDataView  =mlContext.Data.ReadFromBinary(stream);
            #endregion
            #region "STEP 2: Common data process configuration with pipeline data transformations"

            var dataProcessPipeline = mlContext.Transforms.Concatenate("Features", "fontSize",
                                                                                       "isBold",
                                                                                       "isItalic",
                                                                                       "isUnderLine",
                                                                                       "containsDot",
                                                                                       "containsQuestionMark",
                                                                                       "isAllCaps"
                                                                                        //"pageNo",
                                                                                        //"tagHeight",
                                                                                        //"tagWidth",
                                                                                        //"NumericFeatures"
                                                                                        //"fontColor",
                                                                                        //"tagText",
                                                                                        //"firstWord"
                                                                                        );
            var dataProcessPipeline1 = mlContext.Transforms.Text.FeaturizeText(new List<string> { "tagText", "firstWord" }, "Features");
            //var dataProcessPipeline2 = mlContext.Transforms.Text.FeaturizeText(new List<string> { "Label" }, "Features");

            //mlContext.Transforms.CustomMapping()
            #endregion
            #region  "STEP 3: Set the training algorithm, then create and config the modelBuilder"                            
            var modelBuilder = new ModelBuilder<NormalTagsModelFeatures, NormalTagsPrediction>(mlContext, dataProcessPipeline);
            // We apply our selected Trainer 
            modelBuilder.AddEstimator(dataProcessPipeline1);
            //modelBuilder.AddEstimator(new Dictionarizer("Label"));
            //modelBuilder.AddEstimator(dataProcessPipeline2);

            var pipeline = mlContext.Transforms.Conversion.MapValueToKey("Label")
                .Append(dataProcessPipeline1)
                .Append(mlContext.MulticlassClassification.Trainers.StochasticDualCoordinateAscent(labelColumn: "Label", featureColumn: "Features"))
                .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel"));

            //var trainer = mlContext.MulticlassClassification.Trainers.StochasticDualCoordinateAscent(labelColumn: "Label", featureColumn: "Features");
            //modelBuilder.AddTrainer(trainer);
            modelBuilder.AddTrainer(pipeline);
            #endregion
            #region "STEP 4: Train the model fitting to the DataSet"
            //The pipeline is trained on the dataset that has been loaded and transformed.
            //Console.WriteLine("=============== Training the model ===============");

            modelBuilder.Train(trainingDataView);
            #endregion
            #region "STEP 5: Evaluate the model and show accuracy stats"
            //Console.WriteLine("===== Evaluating Model's accuracy with Test data =====");
            var metrics = modelBuilder.EvaluateMultiClassClassificationModel(trainingDataView, "Label");
            //Common.ConsoleHelper.PrintMultiClassClassificationMetrics(trainer.ToString(), metrics);
            #endregion
            #region "STEP 6: Save/persist the trained model to a .ZIP file
            //Console.WriteLine("=============== Saving the model to a file ===============");
            string fullFilePath = GetModelPath("Norm");
            UpdateNormalModelPath(fullFilePath);
            modelBuilder.SaveModelAsFile(fullFilePath);
            #endregion
        }
`
This my Sample Code
